promql/engine.go

```diff
@@ -383,84 +388,55 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *EvalStmt) (
 
 		evalTimer.Stop()
 		ng.metrics.queryInnerEval.Observe(evalTimer.ElapsedTime().Seconds())
-		// Point might have a different timestamp, force it to the evaluation
-		// timestamp as that is when we ran the evaluation.
-		switch v := val.(type) {
-		case Scalar:
-			v.T = start
-		case Vector:
-			for i := range v {
-				v[i].Point.T = start
-			}
-		}
-
-		return val, nil
-	}
-	numSteps := int(s.End.Sub(s.Start) / s.Interval)
 
-	// Range evaluation.
-	Seriess := map[uint64]Series{}
-	for ts := s.Start; !ts.After(s.End); ts = ts.Add(s.Interval) {
-
-		if err := contextDone(ctx, "range evaluation"); err != nil {
-			return nil, err
-		}
-
-		t := timeMilliseconds(ts)
-		evaluator := &evaluator{
-			Timestamp: t,
-			ctx:       ctx,
-			logger:    ng.logger,
-		}
-		val, err := evaluator.Eval(s.Expr)
-		if err != nil {
-			return nil, err
+		mat, ok := val.(Matrix)
+		if !ok {
+			panic(fmt.Errorf("promql.Engine.exec: invalid expression type %q", val.Type()))
 		}
-
-		switch v := val.(type) {
-		case Scalar:
-			// As the expression type does not change we can safely default to 0
-			// as the fingerprint for Scalar expressions.
-			ss, ok := Seriess[0]
-			if !ok {
-				ss = Series{Points: make([]Point, 0, numSteps)}
-				Seriess[0] = ss
-			}
-			ss.Points = append(ss.Points, Point{V: v.V, T: t})
-			Seriess[0] = ss
-		case Vector:
-			for _, sample := range v {
-				h := sample.Metric.Hash()
-				ss, ok := Seriess[h]
-				if !ok {
-					ss = Series{
-						Metric: sample.Metric,
-						Points: make([]Point, 0, numSteps),
-					}
-					Seriess[h] = ss
-				}
-				sample.Point.T = t
-				ss.Points = append(ss.Points, sample.Point)
-				Seriess[h] = ss
+		query.matrix = mat
+		switch s.Expr.Type() {
+		case ValueTypeVector:
+			// Convert matrix with one value per series into vector.
+			vector := make(Vector, len(mat))
+			for i, s := range mat {
+				// Point might have a different timestamp, force it to the evaluation
+				// timestamp as that is when we ran the evaluation.
+				vector[i] = Sample{Metric: s.Metric, Point: Point{V: s.Points[0].V, T: start}}
 			}
+			return vector, nil
+		case ValueTypeScalar:
+			return Scalar{V: mat[0].Points[0].V, T: start}, nil
+		case ValueTypeMatrix:
+			return mat, nil
 		default:
-			panic(fmt.Errorf("promql.Engine.exec: invalid expression type %q", val.Type()))
+			panic(fmt.Errorf("promql.Engine.exec: unexpected expression type %q", s.Expr.Type()))
 		}
+
+	}
+	// Range evaluation.
+	if err := contextDone(ctx, "range evaluation"); err != nil {
```

Comment:
is it worth checking for the context here since it will be checked in `evaluator.Eval` right after?

Reply:
Done.

There's more work to be done on checking for a timeout, but I left it out of this PR.